### DISCUSSION TITLE
Creating a C library

### DISCUSSION THREAD

**Discussion by benstigsen** [zuri-lang/zuri#155] [Q&A] [answered]
So I've been messing around with SDL2, [tigr](https://github.com/erkkah/tigr) and [kit](https://github.com/rxi/kit).

What I've done so far is just take those libraries and compile them like:
```
gcc -lgdi32 -lwinmm -luser32 -c -o tigr.o tigr.c
gcc -lgdi32 -lwinmm -luser32 -shared -o tigr.dll tigr.o
```
And while this does create a `.dll` file, I am wondering if there's a better way to it.

My test library currently looks like so:
```js
var lib = clib.load('./kit.dll')

var kit_create = lib.define('kit_create', clib.ptr, clib.char_ptr, clib.int, clib.int, clib.int)
def create(title, w, h) {
  return kit_create(title, w, h, 0)
}

var kit_step = lib.define('kit_step', clib.bool, clib.ptr, clib.ptr)
def step(ctx) {
  return kit_step(ctx, clib.double)
}

var kit_destroy = lib.define('kit_destroy', clib.void, clib.ptr)
def destroy(ctx) {
  kit_destroy(ctx)
}
```
There's currently some issues with using structs, and it also seems to run inconsistently. What I was thinking is, would it be better if I made a wrapper library in C, using `blade.h`? Should I just create a `.dll` from this or would I have to build it into the Blade executable?

**Comment by mcfriend99** [comment]
Have you checked out the [blade-ext-demo](https://github.com/blade-lang/blade-ext-demo) repository?? That's an example repository showing how to achieve this easily. Though the project uses cmake, it's not exactly a requirement. The basic requirements are:

- Dynamically link against the `libblade`
- Add the `<blade_install_path>/includes` to your includes path

You can learn more about the API by looking at the builtin-packages under the `packages` directory or `src/standard` directory of Blade source code. It's very similar to the builtin methods API as well.

**Comment by benstigsen** [comment]
Yeah but does this build an entirely new Blade executable or DLL with those libraries?
Because ideally I wouldn't need to modify basic Blade installation, for someone to use SDL2 or something.

**Comment by mcfriend99** [comment]
No. It generates a Blade extension dll. If you create a directory `.blade/bin` at the root of the application and run it, it will load it as an extension that can then be used in the project.

For example, if you build the demo project, it will create a `demo.dll` (or `demo.so` for Linux OS and `demo.dylib` on MacOS) file for you. If you put it in a directory called `.blade/bin`, you will be able to import it like this in your application.

```py
import demo

demo.say_hello()
```

You should be greeted by `Hello, world!`.

Also, note that the `sqlite` based message in the CMakeLists.txt file is irrelevant and is no longer valid.

**Reply by benstigsen** [reply]
Ah okay. Well, in that case. I don't know how to compile the `blade-ext-demo` if I follow the instructions:
```
cmake -B ./build -DCMAKE_BUILD_TYPE=Release -DBLADE_ROOT="D:/Projects/Blade/blade/build/blade"
cmake --build build --config Release
```
I've tried changing the path to forward slashes and to just `D:\Projects\Blade\blade` instead of the build, but I keep getting errors.

So I get:
```
CMake Error at CMakeLists.txt:3 (project):
   'nmake' '-?'

  failed with:
   The system cannot find the file specified
```

I also got errors with C compilers, but I defined them in the CMake file:
```
set(CMAKE_CXX_COMPILER "C:/mingw64/bin/g++")
set(CMAKE_C_COMPILER "C:/mingw64/bin/gcc")
```

So now I'm just getting the `The system cannot find the file specified` error.

---

> If you put it in a directory called .blade/bin, you will be able to import it like this in your application.

Nice, I didn't know about this!

**Reply by mcfriend99** [reply]
For the last comment, See https://bladelang.com/tutorial/modules.html#first-class-package-management

**Comment by mcfriend99** [comment, marked as answer]
Do you have the `mingw32-make` in your path?? If you don't, you can get it by installing TCC or WinLibs compiler and adding it to your path (Or simply add the directory `C:/mingw64/bin/` to your path based on what I see). Without changing anything in the CMakeLists.txt file and provided that the `blade.exe` resides in the directory `D:/Projects/Blade/blade/build/blade`, you can change the first command to:

```
cmake -B ./build -DCMAKE_BUILD_TYPE=Release -DBLADE_ROOT="D:/Projects/Blade/blade/build/blade" -DCMAKE_MAKE_PROGRAM=mingw32-make -G "Unix Makefiles"
```

**Comment by benstigsen** [comment]
Alright, your make suggestion made it work! Here's a GIF of a window opening with a loop!

![kit](https://user-images.githubusercontent.com/24865450/228667691-14f7bcdf-fceb-4002-b5d9-202c3fdd45dd.gif)

But, I'm having my hard time wrapping my brain around, how to make it a Blade module. Do I just keep the DLL files in the `.blade/bin`, then what do I do with my `index.b`, `draw.b` and so on? Do I move these into `.blade/libs` and import `_kit`. What would be the best way to make something easily installable or perhaps Nyssa compatible with a C module?

It might be a vague question. But even though I've looked at the `sqlite` and `json` modules, I don't quite see how I could use the same system as those, since they don't make use of a `.blade/` folder.

**Comment by mcfriend99** [comment]
Your suggestion works perfectly. You can convert it into a `nyssa` package and use the nyssa `post_installation` script option to copy the required dlls into the proper directory. 

- [More information about the post_install](https://nyssa.bladelang.com/docs/install-and-uninstall-actions)
- [Sample package using the post_install](https://nyssa.bladelang.com/view/qi)

**Comment by mcfriend99** [comment]
Quick suggestion, you could use `index.b` instead of `main.b` to make it easy to import the package. It's also easier to run with the command `blade .` that way.

Also, which editor is this and how were you able to make it recognise Blade syntax?? I'm really interested.

**Reply by benstigsen** [reply]
I'm using [lite-xl](https://github.com/lite-xl/lite-xl) for writing code in Blade. I added the Blade syntax highlighting to the [plugins](https://github.com/lite-xl/lite-xl-plugins) repo as `language_blade`.

**Comment by benstigsen** [comment]
Sort of related last thing.
Is there any thing gamedev wise you wish was a library for Blade? Because while Kit is good, it's only for Windows. So perhaps you'd prefer Tigr, SDL2 or something else?

**Reply by mcfriend99** [reply]
Tiger is fine, but still looks buggy on osx on resize. I think SDL2 is fine.

**Reply by mcfriend99** [reply]
Is the project on GitHub??

**Reply by benstigsen** [reply]
Not yet. I'll post here when I make it public (within a week).
I also still have to add the HTTP router module to Nyssa.

**Reply by benstigsen** [reply]
I sort of changed my mind. I'm now considering `kit` or `raylib`. Still not certain which one I'll choose. But here's little example with raylib 😄 

![raylib](https://user-images.githubusercontent.com/24865450/228952086-5c74f6fd-d991-40c4-bcd5-4c5faab08b2d.gif)

**Reply by mcfriend99** [reply]
I believe raylib is more matured and cross platform. I think it's a better choice and better suited to Blade as a package.

**Reply by benstigsen** [reply]
Yep, the API is very simple and will fit in nicely with Blade.

Raylib it is!

**Comment by mcfriend99** [comment]
> So I've been messing around with SDL2, [tigr](https://github.com/erkkah/tigr) and [kit](https://github.com/rxi/kit).
> 
> What I've done so far is just take those libraries and compile them like:
> 
> ```
> gcc -lgdi32 -lwinmm -luser32 -c -o tigr.o tigr.c
> gcc -lgdi32 -lwinmm -luser32 -shared -o tigr.dll tigr.o
> ```
> 
> And while this does create a `.dll` file, I am wondering if there's a better way to it.
> 
> My test library currently looks like so:
> 
> ```js
> var lib = clib.load('./kit.dll')
> 
> var kit_create = lib.define('kit_create', clib.ptr, clib.char_ptr, clib.int, clib.int, clib.int)
> def create(title, w, h) {
>   return kit_create(title, w, h, 0)
> }
> 
> var kit_step = lib.define('kit_step', clib.bool, clib.ptr, clib.ptr)
> def step(ctx) {
>   return kit_step(ctx, clib.double)
> }
> 
> var kit_destroy = lib.define('kit_destroy', clib.void, clib.ptr)
> def destroy(ctx) {
>   kit_destroy(ctx)
> }
> ```
> 
> There's currently some issues with using structs, and it also seems to run inconsistently. What I was thinking is, would it be better if I made a wrapper library in C, using `blade.h`? Should I just create a `.dll` from this or would I have to build it into the Blade executable?

This came to my attention again specifically the statement "`There's currently some issues with using structs`". If you knew the elements in the structure and their data type, you can use struct packing available in the `struct` module to create the struct pointer and pass it in as `clib.ptr`. Blade struct packing behaves the same as and uses the same syntax as `PHP`'s struct packing.

See  the [structs](https://bladelang.com/standard/struct.html) documentation.

**Comment by benstigsen** [comment]
I've been busy with school and work so I haven't gotten around to the raylib library for Blade. But I can see that you have started Blade projects based on raylib and raygui for these 😄

Do you still want me to make a 1:1 raylib library?

**Reply by mcfriend99** [reply]
More of an experiment and a way to test the clib module ahead of v0.0.84 and not something I'm planning to release. Will be taking down the GitHub repo as soon as th clib module matures.
